### ISSUE TITLE
[Question] Getting the sha of a commit closing an issue

### ISSUE THREAD

**Issue by stefanodallapalma** [python-gitlab/python-gitlab#1208]
Hi,
given a closed issue, I would like to retrieve the commit that closed it. 

For example, in https://gitlab.com/NickBusey/HomelabOS/-/issues/568 there is a list of events (from oldest to newest):

-  Nick Busey @NickBusey created merge request **!589** (closed) to address this issue 3 months ago
...
-  Nick Busey @NickBusey closed via commit **3e976934** 3 months ago

I would like to identify the commit [3e9769349dd8b98e9d35a861b2913eafafa1e89b](https://gitlab.com/NickBusey/HomelabOS/-/commit/3e9769349dd8b98e9d35a861b2913eafafa1e89b) through the APIs. 
So far, I have been able to identify the sha (a8cd2f47c3b6a6ad50fc2e5bc853540bee5ae6e4) of the [merging request](https://gitlab.com/api/v4/projects/6853087/merge_requests/568) only. 

This is my snippet:

```
gl = gitlab.Gitlab('http://gitlab.com', 'access_token')
project = gl.projects.get('NickBusey/HomelabOS')
closed_issues = project.issues.list(state='closed', labels=['Bug'], all=True)
for issue in closed_issues:
    mr = project.mergerequests.get(issue.iid)
    print(issue.iid, mr.sha)
```

I tried with the Events APIs as well, but I am not able to find that information anywhere.

Any help?

Regards,
Stefano

## Specifications
  - python-gitlab version: 2.5.0

**Comment by nejch** [comment]
Hi @stefanodallapalma, I had a quick look through relevant pages at https://docs.gitlab.com/ee/api/ and think this is something that the GitLab API does not expose, so python-gitlab will not be able to provide it either. But please re-open this if you find it there.

However, GitLab simply does this with the "Auto-close referenced issues on default branch." So I think you can get a bit creative and do it the other way around: by listing commits on the default branch, then matching the closing pattern of their commit messages against your issue ID. See https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically for details on the closing patterns etc.

Something like this (totally untested, might not even be valid syntax 😁 but just a pseudo-code idea):
```python
import re

def _matches_closing_pattern(commit, issue):
    closing_pattern = re.compile('..') # Add real pattern from link above with issue.iid
    if re.search(closing_pattern, commit.message):
        return True

    return False

# ... rest of your code

# Probably change as needed, https://python-gitlab.readthedocs.io/en/stable/gl_objects/commits.html
commits = project.commits.list(ref_name=project.default_branch)

for issue in closed_issues:
    try:
        closing_commit = [commit.sha for commit in commits if _matches_closing_pattern(commit, issue][0]
    except IndexError:
        print(f"No closing commits found for issue {issue.iid}.")
````

See if you can come up with a working example and share the snippet with us here 😃

**Comment by stefanodallapalma** [comment]
Hi @nejch,

thanks for the reply, I really appreciate it. I came with a different solution though.
Iterating over commits for every closed issue might be unproductive, especially in cases where the project has thousands of commits and hundreds of closed issues.

Below my solution, which uses the [Notes API](https://docs.gitlab.com/ee/api/notes.html#list-all-merge-request-notes).

```
import re
commit_sha_pattern = re.compile('closed via commit ([0-9a-f]{5,40})')

closed_issues = project.issues.list(state='closed', labels=['Bug'], all=True)
for issue in closed_issues:
    for note in issue.notes.list(all=True, as_list=False, sort='desc'):
        match = re.search(commit_sha_pattern, note.body)
        if match:
            print('Commit closing the issue:', match.groups()[0])
            break
```

I will explain for those that have a similar issue.
In this snippet, I iterate over closed issues. For each issue I get the list of notes (```issue.notes.list(all=True, as_list=False, sort='desc')```), like:

* Nick Busey @NickBusey created merge request !589 (closed) to address this issue
* Nick Busey @NickBusey closed via commit 3e976934

whose are returned in the python as follow:

```
<class 'gitlab.v4.objects.ProjectIssueNote'> => {'id': 375977497, 'body': 'created merge request !589 (closed) to address this issue', ...}
<class 'gitlab.v4.objects.ProjectIssueNote'> => {'id': 375977496, 'body': 'closed via commit 3e9769349dd8b98e9d35a861b2913eafafa1e89b', ...}
```

Then, ```match = re.search(commit_sha_pattern, note.body)``` tries to match the pattern "closed via commit <sha>". If a match is found, then access the found group, that is ```match.groups()[0]``` based on the pattern ```commit_sha_pattern```, and go to the next issue (```break```).
 
**Note: ** I iterated over notes using the parameters:
* ```as_list=False``` to return a Generator instead of a list (saving a little of time, since I also used the break construct), and 
* ```sort='desc'```: to iterate starting since the newest "event" until the oldest. In this case, I stop when the regex matches the newest commit that closed the issue (it might happen that an issue is re-opened several time and closed as many time by a commit).

**Comment by stefanodallapalma** [comment]
I've noticed that sometimes the note: "closed by commit <sha>" is not present,  because the commit does not close the issue directly, but the merge request does.
In this case, a different solution would be to look for eitehr notes (1) "closed via merge request !<IID>" and (2) "closed via commit <sha>".
Then, if (2) is matched, get the sha directly, otherwise, call the [merge_requests API](https://docs.gitlab.com/ee/api/merge_requests.html) and get the associated sha.

```
import re
closed_via_commit = re.compile('closed via commit ([0-9a-f]{5,40})')
closed_via_mr = re.compile('closed via merge request !([0-9]+)')

closed_issues = project.issues.list(state='closed', labels=['Bug'], all=True)
i = 1
for issue in closed_issues:
    mr_iid = None
    sha = None

    for note in issue.notes.list(all=True, as_list=False, sort='desc'):
        match = re.search(closed_via_mr, note.body)
        if match:
            mr_iid = match.groups()[0]
            # still look for commit sha
    
        match = re.search(closed_via_commit, note.body)
        if match:
            sha = match.groups()[0]
            break
    
    if sha:
        print('Closed by commit', sha)
    elif mr_iid:
        mr = project.mergerequests.get(mr_iid)
        print('Closed by commit', mr.sha, 'in merge request', mr_iid)
```
